## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [11]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [12]:
# View the column names in the data
print(df.columns)


Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')


In [13]:
# Use the describe function to gather some basic statistics
description = df.describe()
description


client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [14]:
# looking at top 10.
df.head(10)

first      last                                           job  \
0     Donald   Harding                                  Immunologist   
1    Tiffany     Myers                               Music therapist   
2    Shannon    Watson                                  Immunologist   
3     Nathan     Baker                         Accounting technician   
4  Christina  Schwartz                                  Chiropractor   
5    Jessica     Reyes  Production designer, theatre/television/film   
6      Brian     Moore                                       Barista   
7      Jesus    Wright                            Theme park manager   
8    Rebecca     Brown                                   Chiropodist   
9       Greg    Garcia                            Secretary, company   

                 phone                                email  client_id  \
0   793-904-7725x39308     harding.donald.7185@sullivan.com      58515   
1     201.442.4543x942    myers.t.6537@ferguson-johnson.net      37609   
2    687.737.9424x8503                swatson8146@payne.net      57113   
3     827-788-8123x012               bakernathan@benson.com      46554   
4         265-829-3643  christinaschwartz9252@mcconnell.com      92089   
5           5016915923             jessica.reyes@grimes.net      33615   
6  +1-893-400-2733x982          moore.official@martinez.com      87213   
7   605.947.9467x66513        wright.jesus.7579@meadows.com      46709   
8     665-955-0156x926                 rbrown850@lewis.info      73362   
9   (231)342-7216x1421                  ggarcia@bentley.biz      58789   

   order_id  order_date  order_week  order_year         item_id     category  \
0   8953482  2023-04-28          17        2023  EUD29711-63-6U        decor   
1   8069089  2023-05-19          20        2023  XDA18116-89-4A  consumables   
2   1902144  2023-01-29           4        2023  ABE59463-05-7E     software   
3   9031802  2023-04-25          17        2023  ZMM00836-65-0C  consumables   
4   1322274  2023-05-28          21        2023  BZX55559-12-3X  consumables   
5   1327730  2023-05-02          18        2023  BMB23158-88-9A        decor   
6   8080280  2023-04-19          16        2023  UAC50632-25-4E  consumables   
7   8946782  2023-02-14           7        2023  UEU25316-01-1U  consumables   
8   9394118  2023-05-13          19        2023  DDA06248-98-7Z  consumables   
9   9665852  2023-04-14          15        2023  DUU85042-87-5E    furniture   

          subcategory  unit_price  unit_cost  unit_weight  qty  line_number  
0            wall art     1096.80     762.71         7.50  105            1  
1                pens       24.95      15.09         1.49   21            0  
2  project management       13.52       7.86         1.68   39            6  
3                pens       36.42      24.85         1.23   29            3  
4                misc      195.10     108.17        46.43   20            1  
5              plants      172.84     154.44         2.79   78            0  
6    kitchen supplies      245.56     231.95         1.73   56            5  
7                pens       20.99      16.97         5.12   22            4  
8      paper products       83.86      59.00         0.27   49            3  
9             seating       66.88      54.15         2.47   78            1

In [15]:
# Look at last 10.
df.tail(10)

first       last                                      job  \
54629    Linda  Henderson  Scientist, research (physical sciences)   
54630     Lisa     Riddle          Chartered management accountant   
54631    Kevin     Obrien                           Microbiologist   
54632   Austin     Fuller                       Investment analyst   
54633   Tamara      Jones                        Fish farm manager   
54634   Lauren      Reese                           Radio producer   
54635  Derrick      Moore                                 Musician   
54636   Monica  Gutierrez                         Graphic designer   
54637    Wanda    Solomon                             Toxicologist   
54638   Dwayne       Bell                  Chief Financial Officer   

                     phone                                email  client_id  \
54629  +1-664-405-4100x443         lhenderson@boyer-krueger.com      79817   
54630     001-932-725-9686        riddlel1336@hamilton-cole.net      19904   
54631           7112448766         line.kevin@hart-williams.com      39939   
54632         379.662.7549           fulleraustin7494@nunez.com      50317   
54633     375.495.3124x931  tamara.jones.9326@george-carter.com      22109   
54634   977.877.5272x11382   late_reese_4081@montoya-chavez.org      29043   
54635         358.661.5483         derrick.moore.2602@pope.info      41908   
54636     294.805.9100x339            gutierrezm3195@morris.org      35176   
54637        (311)767-4924            solomonwanda5962@ross.org      24485   
54638    606-619-7667x9508                     belld@galvan.com      24007   

       order_id  order_date  order_week  order_year         item_id  \
54629   6473704  2023-03-16          11        2023  ZCC45621-74-4A   
54630   6871726  2023-03-18          11        2023  CCB77138-94-1M   
54631   6430193  2023-04-07          14        2023  EAA51108-37-5D   
54632   5064968  2023-05-26          21        2023  CUE68543-95-3U   
54633   8568920  2023-04-26          17        2023  CDX06274-12-1E   
54634   9021716  2023-04-26          17        2023  UCA28532-31-0B   
54635   6290153  2023-02-08           6        2023  EBD29277-23-7U   
54636   8692622  2023-03-05           9        2023  EBD30093-55-9Z   
54637   7592730  2023-01-18           3        2023  XMZ38239-49-3D   
54638   7489403  2023-01-04           1        2023  ZCZ05589-50-5D   

          category         subcategory  unit_price  unit_cost  unit_weight  \
54629  consumables    kitchen supplies       72.60      52.42         0.33   
54630     software           logistics      454.34     316.29         1.17   
54631    furniture             seating       12.11       8.62         1.45   
54632  consumables      paper products      557.31     409.72         2.21   
54633  consumables    kitchen supplies        8.16       5.10         3.84   
54634     software           logistics       83.13      51.60         2.25   
54635  consumables    kitchen supplies      206.59     175.46        11.70   
54636  consumables   bathroom supplies       65.66      57.31         4.16   
54637    furniture              stands        1.48       1.22        18.04   
54638     software  project management        3.01       1.61         2.07   

        qty  line_number  
54629   153            1  
54630    48            7  
54631    70            1  
54632   330            1  
54633    41            2  
54634    33            8  
54635    47            0  
54636   475            6  
54637   112            0  
54638  1031            5

In [16]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df.groupby('order_id')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [17]:
# What three item categories had the most entries?
most entries?
top_categories = df['category'].value_counts().head(3)

top_categories


Object `entries` not found.


category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [18]:
# For the category with the most entries,
# which subcategory had the most entries?
top_subcategories = df['subcategory'].value_counts().head(3)
top_subcategories


subcategory
bathroom supplies    6424
paper products       5874
kitchen supplies     5746
Name: count, dtype: int64

In [19]:
# Which five clients had the most entries in the data?
top_5_clients = df['client_id'].value_counts().head(5)

print(top_5_clients)


client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64


In [23]:
# Store the client ids of those top 5 clients in a list.
top_5_clients_ids = df['client_id'].value_counts().head(5)

print(top_5_clients_ids)


client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64


In [24]:
# How many total units (the qty column) did the
# client with the most entries order order?
top_client = df['client_id'].value_counts().idxmax()

# Calculate the total units ordered by this client
total_units = df[df['client_id'] == top_client]['qty'].sum()

print(f"The client with the most entries ({top_client}) ordered a total of {total_units} units.")


The client with the most entries (33615) ordered a total of 64313 units.


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [25]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty
df['subtotal'] = df['unit_price'] * df['qty']

# Display the first few rows of the DataFrame to verify the new column
print(df[['unit_price', 'qty', 'subtotal']].head())


   unit_price  qty   subtotal
0     1096.80  105  115164.00
1       24.95   21     523.95
2       13.52   39     527.28
3       36.42   29    1056.18
4      195.10   20    3902.00


In [26]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.
# Create shipping_price column
df['total_weight'] = df['unit_weight'] * df['qty']
df['shipping_price'] = df['total_weight'].apply(lambda x: x * 7 if x > 50 else x * 10)

# Create subtotal column
df['subtotal'] = df['unit_price'] * df['qty']

# Display the first few rows to verify the new columns
print(df[['unit_weight', 'qty', 'total_weight', 'shipping_price', 'unit_price', 'subtotal']].head())


   unit_weight  qty  total_weight  shipping_price  unit_price   subtotal
0         7.50  105        787.50         5512.50     1096.80  115164.00
1         1.49   21         31.29          312.90       24.95     523.95
2         1.68   39         65.52          458.64       13.52     527.28
3         1.23   29         35.67          356.70       36.42    1056.18
4        46.43   20        928.60         6500.20      195.10    3902.00


In [27]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%

# Define the sales tax rate
SALES_TAX_RATE = 0.0925  # 9.25%

# Calculate the total price
df['total_price'] = (df['subtotal'] + df['shipping_price']) * (1 + SALES_TAX_RATE)

# Display the first few rows to verify the new column
print(df[['subtotal', 'shipping_price', 'total_price']].head())



    subtotal  shipping_price    total_price
0  115164.00         5512.50  131839.076250
1     523.95          312.90     914.258625
2     527.28          458.64    1077.117600
3    1056.18          356.70    1543.571400
4    3902.00         6500.20   11364.403500


In [28]:
# Create a column for the cost

# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).
# Calculate the cost for each line
df['line_cost'] = (df['unit_cost'] * df['qty']) + df['shipping_price']

# Display the first few rows to verify the new column
print(df[['unit_cost', 'qty', 'shipping_price', 'line_cost']].head())


   unit_cost  qty  shipping_price  line_cost
0     762.71  105         5512.50   85597.05
1      15.09   21          312.90     629.79
2       7.86   39          458.64     765.18
3      24.85   29          356.70    1077.35
4     108.17   20         6500.20    8663.60


In [29]:
# Create a column for the profit of
# each line using line cost and line price
# Calculate the profit for each line
df['line_profit'] = (df['subtotal'] + df['shipping_price']) - df['line_cost']

# Display the first few rows to verify the new column
print(df[['subtotal', 'shipping_price', 'line_cost', 'line_profit']].head())


    subtotal  shipping_price  line_cost  line_profit
0  115164.00         5512.50   85597.05     35079.45
1     523.95          312.90     629.79       207.06
2     527.28          458.64     765.18       220.74
3    1056.18          356.70    1077.35       335.53
4    3902.00         6500.20    8663.60      1738.60


## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [30]:
# Check your work using the totals above
# Function to check an order
def check_order(order_id, expected_total):
    order_total = df[df['order_id'] == order_id]['total_price'].sum()
    print(f"Order ID {order_id}:")
    print(f"Calculated total: ${order_total:.2f}")
    print(f"Expected total:   ${expected_total:.2f}")
    print(f"Difference:       ${order_total - expected_total:.2f}")
    print(f"Match: {abs(order_total - expected_total) < 0.01}\n")  # Allow for small floating point differences

# Check each order
check_order(2742071, 152811.89)
check_order(2173913, 162388.71)
check_order(6128929, 923441.25)

Order ID 2742071:
Calculated total: $152811.90
Expected total:   $152811.89
Difference:       $0.01
Match: True

Order ID 2173913:
Calculated total: $162388.72
Expected total:   $162388.71
Difference:       $0.01
Match: True

Order ID 6128929:
Calculated total: $923441.24
Expected total:   $923441.25
Difference:       $-0.01
Match: True



## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [31]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.
# Step 1: Identify the top 5 clients by quantity
top_5_clients = df.groupby('client_id')['qty'].sum().nlargest(5)

print("Top 5 clients by quantity:")
print(top_5_clients)
print()

# Step 2: Calculate total spending for each of these clients
for client_id in top_5_clients.index:
    client_spending = df[df['client_id'] == client_id]['total_price'].sum()
    print(f"Client ID {client_id} spent: ${client_spending:.2f}")


Top 5 clients by quantity:
client_id
78965    4033302
86721    1010426
30143     846069
85931     644423
77354     506382
Name: qty, dtype: int64

Client ID 78965 spent: $655923670.06
Client ID 86721 spent: $181266297.07
Client ID 30143 spent: $33328849.83
Client ID 85931 spent: $237417001.42
Client ID 77354 spent: $71942651.78


In [32]:
# Create a summary DataFrame showing the totals for the
# for the top 5 clients with the following information:
# total units purchased, total shipping price,
# total revenue, and total profit. Sort by total profit.

# Identify the top 5 clients by quantity
top_5_clients = df.groupby('client_id')['qty'].sum().nlargest(5).index

# Create the summary DataFrame
summary = df[df['client_id'].isin(top_5_clients)].groupby('client_id').agg({
    'qty': 'sum',
    'shipping_price': 'sum',
    'total_price': 'sum',
    'line_profit': 'sum'
})

# Rename columns for clarity
summary.columns = ['Total Units', 'Total Shipping', 'Total Revenue', 'Total Profit']

# Sort by Total Profit
summary_sorted = summary.sort_values('Total Profit', ascending=False)

# Display the summary
print(summary_sorted)

           Total Units  Total Shipping  Total Revenue  Total Profit
client_id                                                          
78965          4033302      5340756.88   6.559237e+08  1.918170e+08
85931           644423     10122549.19   2.374170e+08  8.919246e+07
86721          1010426     64335727.72   1.812663e+08  2.762721e+07
30143           846069      8981905.24   3.332885e+07  8.428492e+06
77354           506382     23814685.86   7.194265e+07  6.341006e+06


In [33]:
# Format the data and rename the columns
# to names suitable for presentation.
# Currency should be in millions of dollars.

# Convert currency columns to millions of dollars
currency_columns = ['Total Revenue ($ millions)', 'Total Profit ($ millions)']
df[currency_columns] = df[currency_columns] / 1e6

# Convert Total Shipping Cost to dollars (it's already in dollars)
df['Total Shipping Cost ($)'] = df['Total Shipping Cost ($)'].round(2)

# Sort the DataFrame by 'Total Profit ($ millions)' in descending order
df_sorted = df.sort_values(by='Total Profit ($ millions)', ascending=False)

# Round all currency columns to 2 decimal places
df_sorted[currency_columns + ['Total Shipping Cost ($)']] = df_sorted[currency_columns + ['Total Shipping Cost ($)']].round(2)

# Display the sorted and formatted DataFrame
print(df_sorted)


KeyError: "None of [Index(['Total Revenue ($ millions)', 'Total Profit ($ millions)'], dtype='object')] are in the [columns]"

In [10]:
# Look at last 10.
df.tail(10)

NameError: name 'df' is not defined

In [ ]:
# Sort the updated data by "Total Profit" form highest to lowest
df_sorted[currency_columns] = df_sorted[currency_columns].round(2)

# Display the sorted DataFrame
print(df_sorted)

In [6]:
# Define the currency columns
currency_columns = ['Unit Price ($ millions)', 'Unit Cost ($ millions)', 'Total Profit ($ millions)']

# Sort the DataFrame by 'Total Profit ($ millions)' in descending order
df_sorted = df.sort_values(by='Total Profit ($ millions)', ascending=False)

# Round the currency columns to 2 decimal places
df_sorted[currency_columns] = df_sorted[currency_columns].round(2)

# Display the sorted DataFrame
print(df_sorted)

NameError: name 'df' is not defined